src/librustc_mir/interpret/memory.rs

```diff
@@ -1236,17 +1234,17 @@ pub(crate) trait HasDataLayout<'a> : Copy {
     }
 }
 
-impl<'a> HasDataLayout<'a> for &'a TargetDataLayout {
+impl<T: layout::HasDataLayout> PointerArithmetic for T {}
+
+impl<'a, 'tcx> layout::HasDataLayout for &'a Memory<'a, 'tcx> {
     #[inline]
-    fn data_layout(self) -> &'a TargetDataLayout {
-        self
+    fn data_layout(&self) -> &TargetDataLayout {
```

Comment:
Ah, that's the difference to what I tried -- you are passing `&&Memory` here, which seems a little weird. Whatever^^